app/models/concerns/stat/activity_history.rb

```diff
@@ -2,59 +2,212 @@ class Stat < ApplicationRecord
-      'total' => 0,
-      'activity' => []
+      'days' => {},
+      'first_event_date' => 5.years.ago,
+      'last_update_date' => 5.years.ago,
+      'week_high_score' => 0,
+      'total_progress' => 0
     }.freeze
 
     def recalculate!
-      activity = user.library_events.by_kind(media_column)
-                     .eager_load(media_column)
-
-      # clear stats_data
       self.stats_data = {}
-      self.stats_data = {
-        total: activity.count,
-        activity: activity # collection of events
-      }
+
+      date_trunc = "date_trunc('day', library_events.created_at)"
+      progress0 = "(changed_data#>>'{progress,0}')::integer"
+      progress1 = "(changed_data#>>'{progress,1}')::integer"
+      progress = "(#{progress1} - #{progress0})"
+
+      case media_column
+      when :anime then watch_time = "sum(#{progress} * anime.episode_length)"
+      when :manga then watch_time = 0
+      end
+
+      activity_dates = user.library_events.where("changed_data ? 'progress'")
+                           .group(date_trunc)
+                           .eager_load(media_column)
+                           .order("#{date_trunc} ASC")
+                           .pluck("#{date_trunc}, sum(#{progress}), #{watch_time}")
+
+      # Set all the days with 0
+      stats_data['days'] = generate_missing_days(activity_dates.first[0])
+      # Set default for total_progress on top level
+      stats_data['total_progress'] = 0
+      # Record the date their first library_event was created
+      stats_data['first_event_date'] = activity_dates.first[0].to_date
+      stats_data['last_update_date'] = activity_dates.last[0].to_date
+      # Set the progress and time for each day
+      # Set the total_progress from all days
+      activity_dates.each do |date, amount, time|
+        stats_data['days'][date.year.to_s][date.month.to_s][date.day.to_s] = {
+          'total_progress' => amount,
+          'total_time' => time
+        }
+        stats_data['total_progress'] += amount
+      end
+      # Set the week_high_score from their first library_event date
+      stats_data['week_high_score'] = week_high_score(activity_dates.first[0])
 
       save!
     end
 
+    def generate_missing_days(start_date)
+      # not happy about having to use Time.now but I think I have to...
+      dates = start_date.to_date..Time.now.to_date
+
+      dates.each_with_object(preset_hash) do |date, h|
+        h[date.year.to_s][date.month.to_s][date.day.to_s] = {
+          'total_progress' => 0,
+          'total_time' => 0
+        }
+      end
+    end
+
+    def find_missing_days(library_event)
+      last_update = stats_data['last_update_date'].to_date + 1.day
+
+      if last_update <= library_event.created_at.to_date
+        stats_data['days'].deep_merge!(record.generate_missing_days(last_update))
+      end
+
+      self
+    end
+
+    def preset_hash
+      Hash.new { |h, y| h[y] = Hash.new { |h2, m| h2[m] = Hash.new { |h3, d| h3[d] = {} } } }
```

Comment:
I think this can be used instead: `Hash.new { |hash, key| hash[key] = Hash.new(&hash.default_proc) }`.

Reply:
don't worry about the activity history stuff, I'm gonna be refactoring it